### PR TITLE
Add Performance Tuning Guide

### PR DESCRIPTION
**Pull request by maximiliankolb**
This PR migrates the "Satellite Performance Tuning Guide" from https://github.com/RedHatSatellite/satellite-performance-tuning.

Discussion: https://github.com/RedHatSatellite/satellite-performance-tuning/issues/33

Cherry-pick into:

- [x] Foreman 3.3
- [x] Foreman 3.2
- [x] Foreman 3.1

### CODE REVIEW

**Comment by maximiliankolb** [comment]
Current status: **draft**; I need to check the migrated content, esp. links, branding, and tables.

**Comment by ekohl** [comment]
> Current status: **draft**; I need to check the migrated content, esp. links, branding, and tables.

When creating a PR you can mark it as draft. The UI has a button, [hub](https://github.com/github/hub) has a `-d` argument. I've now used the Convert to draft button. Feel free to hit `Ready for review` whenever you think it's ready. You can always convert it back to draft.

**Comment by maximiliankolb** [comment]
**DO NOT MERGE**

This PR is ready for a first review. It's not completely done yet, but I need some input and feedback.

**Comment by maximiliankolb** [comment]
Rebased to master. I don't know why it's not building. Any input would be greatly appreciated!

File: guides/common/modules/proc_configuring-the-open-files-limit-for-apache-httpd.adoc
Comment (ekohl) [inline guides/common/modules/proc_configuring-the-open-files-limit-for-apache-httpd.adoc:20]: I hate this if I'm honest. I really don't know why we don't just use
```suggestion
# systemtl restart httpd
```

There is 0 reason to restart all services and only takes a long time.

I think there is also an option like `--only` to foreman-maintain but really, we're already deep into systemd by editing an override file (it doesn't mention you need to create `/etc/systemd/system/httpd.service.d` for example). It also calls daemon-reload.

File: guides/common/modules/proc_configuring-the-qpid-mgmt-pub-interval-parameter.adoc
Comment (ekohl) [inline guides/common/modules/proc_configuring-the-qpid-mgmt-pub-interval-parameter.adoc:21]: This is bad advise since the installer will revert this anyway. Instead, it should advise custom-hiera.yaml:
```yaml
qpid::mgmt_pub_interval: 5
```

Then rerun the installer.

File: guides/common/modules/proc_enabling-tuned-profiles.adoc
Comment (ekohl) [inline guides/common/modules/proc_enabling-tuned-profiles.adoc:6]:
```suggestion
On virtual machines, {Team} recommends to run the `virtual-guest` profile.
```

File: guides/common/modules/proc_enabling-tuned-profiles.adoc
Comment (ekohl) [inline guides/common/modules/proc_enabling-tuned-profiles.adoc:19]: `enable` does not start the service. It just enables it (which means: start it at boot). If you want to start it
```suggestion
. If `tuned` is not running, enable it:
+
[options="nowrap" subs="+quotes,attributes"]
----
# systemctl enable --now tuned
```

File: guides/common/modules/proc_enabling-tuned-profiles.adoc
Comment (ekohl) [inline guides/common/modules/proc_enabling-tuned-profiles.adoc:37]: This isn't just us, it's really a best practice. https://www.percona.com/blog/2019/03/06/settling-the-myth-of-transparent-hugepages-for-databases/ actually has some nice write up.

File: guides/common/modules/con_puma-workers-and-threads-autotuning.adoc
Comment (ekohl) [inline guides/common/modules/con_puma-workers-and-threads-autotuning.adoc:34]: Wouldn't it be easier to see the progress tree using `systemctl status foreman.service`?

**Review by ekohl — COMMENTED** [review]
Not a complete review since I didn't read through the Puma since the recommendations changed.

File: guides/common/modules/con_performance-tuning-quickstart.adoc
Comment (maximiliankolb) [reply guides/common/modules/con_performance-tuning-quickstart.adoc:26]: I've made this visible for ProjectServer only for Katello and downstream.

File: guides/common/modules/con_postgresql-tuning.adoc
Comment (maximiliankolb) [reply guides/common/modules/con_postgresql-tuning.adoc:10]: Yes, added.

File: guides/common/modules/con_postgresql-tuning.adoc
Comment (maximiliankolb) [reply guides/common/modules/con_postgresql-tuning.adoc:8]: Applied.

File: guides/common/modules/con_postgresql-tuning.adoc
Comment (maximiliankolb) [reply guides/common/modules/con_postgresql-tuning.adoc:38]: Removed.

File: guides/common/modules/con_postgresql-tuning.adoc
Comment (maximiliankolb) [reply guides/common/modules/con_postgresql-tuning.adoc:33]: Added.

File: guides/common/modules/con_puma-threads.adoc
Comment (maximiliankolb) [reply guides/common/modules/con_puma-threads.adoc:4]: Applied.

File: guides/common/modules/con_smart-proxy-configuration-tuning.adoc
Comment (maximiliankolb) [reply guides/common/modules/con_smart-proxy-configuration-tuning.adoc:4]: Reworded.

File: guides/common/modules/con_top-performance-considerations.adoc
Comment (maximiliankolb) [reply guides/common/modules/con_top-performance-considerations.adoc:9]: Applied.

File: guides/common/modules/con_top-performance-considerations.adoc
Comment (maximiliankolb) [reply guides/common/modules/con_top-performance-considerations.adoc:14]: Applied.

File: guides/common/modules/proc_benchmark-disk-performance.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_benchmark-disk-performance.adoc:10]: Applied.

File: guides/common/modules/proc_benchmark-disk-performance.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_benchmark-disk-performance.adoc:4]: Added the link (& I agree).

**Comment by maximiliankolb** [comment]
Rebased to master & resolved various issues. Thanks @ekohl for your review. I am still working on addressing all your comments. **Sorry** if you got a Mail each time I've responded in a thread. I forgot that I should be using the "diff" tab and actually conduct a review.

File: guides/common/modules/proc_benchmarking-raw-db-performance.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_benchmarking-raw-db-performance.adoc]: Changed the directory.

File: guides/common/modules/proc_calculating-the-maximum-open-files-limit-for-qdrouterd.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_calculating-the-maximum-open-files-limit-for-qdrouterd.adoc]: Described it as default value.

File: guides/common/modules/proc_calculating-the-maximum-open-files-limit-for-qdrouterd.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_calculating-the-maximum-open-files-limit-for-qdrouterd.adoc]: Applied.

File: guides/common/modules/proc_calculating-the-maximum-open-files-limit-for-qpidd.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_calculating-the-maximum-open-files-limit-for-qpidd.adoc]: Fixed.

File: guides/common/modules/proc_calculating-the-maximum-open-files-limit-for-qpidd.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_calculating-the-maximum-open-files-limit-for-qpidd.adoc]: Applied.

File: guides/common/modules/proc_configuring-how-many-processes-can-be-launched-by-apache-httpd.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_configuring-how-many-processes-can-be-launched-by-apache-httpd.adoc:24]: Added.

File: guides/common/modules/proc_configuring-how-many-processes-can-be-launched-by-apache-httpd.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_configuring-how-many-processes-can-be-launched-by-apache-httpd.adoc]: Yes, fixed.

File: guides/common/modules/proc_configuring-how-many-processes-can-be-launched-by-apache-httpd.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_configuring-how-many-processes-can-be-launched-by-apache-httpd.adoc]: Removed.

File: guides/common/modules/proc_configuring-project-environment-for-performance.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_configuring-project-environment-for-performance.adoc]: Applied.

File: guides/common/modules/proc_configuring-project-environment-for-performance.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_configuring-project-environment-for-performance.adoc:14]: Added.

File: guides/common/modules/proc_configuring-the-maximum-asynchronous-input-output-requests.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_configuring-the-maximum-asynchronous-input-output-requests.adoc]: Applied.

File: guides/common/modules/proc_configuring-the-open-files-limit-for-apache-httpd.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_configuring-the-open-files-limit-for-apache-httpd.adoc]: Applied.

File: guides/common/modules/proc_configuring-the-qpid-mgmt-pub-interval-parameter.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_configuring-the-qpid-mgmt-pub-interval-parameter.adoc]: Thanks, applied!

File: guides/common/modules/proc_enabling-tuned-profiles.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_enabling-tuned-profiles.adoc]: Applied.

File: guides/common/modules/proc_enabling-tuned-profiles.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_enabling-tuned-profiles.adoc]: Applied.

File: guides/common/modules/proc_enabling-tuned-profiles.adoc
Comment (maximiliankolb) [reply guides/common/modules/proc_enabling-tuned-profiles.adoc]: Reworded.

**Review by maximiliankolb — COMMENTED** [review]
Thanks so much for all your input! I've resolved some more issues. I will now try to get it to acutally build.

**Comment by maximiliankolb** [comment]
This PR is now officially ready for review. :) It aims be to a copy of the content from https://github.com/RedHatSatellite/satellite-performance-tuning/ with minor improvements and corrections.

File: guides/common/modules/con_puma-workers-and-threads-recommendation.adoc
Comment (jhutar) [reply guides/common/modules/con_puma-workers-and-threads-recommendation.adoc:15]: Hello Ewoud. That is what we measured - IIRC there was some slight improvement with more threads - we have "Our recommendation is based purely on concurrent registration performance, so it might not reflect your exact use-case." there.

File: guides/common/modules/con_postgresql-tuning.adoc
Comment (jhutar) [reply guides/common/modules/con_postgresql-tuning.adoc:24]: Yep, that is old.

**Comment by jhutar** [comment]
Hello. I agree with Ewoud's comments. Could we maybe merge as it is to "snapshot" initial version of the tuning guide and then work on implementing these proposed changes?

**Comment by maximiliankolb** [comment]
> Hello. I agree with Ewoud's comments. Could we maybe merge as it is to "snapshot" initial version of the tuning guide and then work on implementing these proposed changes?

That would be awesome! I don't really have the knowledge and time to get into every comment Ewoud made; that why some threads are still open. Of course I can assist with and review future PRs for this guide.

File: guides/common/modules/con_postgresql-tuning.adoc
Comment (maximiliankolb) [reply guides/common/modules/con_postgresql-tuning.adoc]: I have claried this option.

File: guides/common/modules/con_postgresql-tuning.adoc
Comment (maximiliankolb) [reply guides/common/modules/con_postgresql-tuning.adoc:34]: I need help on this one. cc @jhutar

File: guides/common/modules/con_puma-threads.adoc
Comment (maximiliankolb) [reply guides/common/modules/con_puma-threads.adoc:16]: Can you confirm @jhutar? Note that we are providing this only for Foreman 3.1+.

File: guides/common/modules/con_puma-workers-and-threads-autotuning.adoc
Comment (maximiliankolb) [reply guides/common/modules/con_puma-workers-and-threads-autotuning.adoc]: Yes, added.

**Review by maximiliankolb — COMMENTED** [review]
I have rebased my branch to master again & fixed a couple more issues. I need some more help for the remaining open threads.

File: guides/common/modules/con_postgresql-tuning.adoc
Comment (jhutar) [reply guides/common/modules/con_postgresql-tuning.adoc:34]: Right, this is very outdated now. What about trimming it to this for now:
```
`max_connections`: The key defines the maximum number of connections that can be accepted by the PostgreSQL processes that are running.
```
I have created https://issues.redhat.com/browse/SAT-10307 to track figuring out more details.

File: guides/common/modules/con_puma-threads.adoc
Comment (jhutar) [reply guides/common/modules/con_puma-threads.adoc:16]: Yep, we also recommend min threads == max threads.

We recommended 16 because we had a better results with it.

I agree tuning Puma workers is more straight forward.

This is what I released yesterday as an update to 6.10 tuning guide: https://github.com/RedHatSatellite/satellite-performance-tuning/commit/b87caa34b409aa36c440d9ef044c9fd8ce964ea0#diff-4d49ae92acb7629a016fdcecfa8cfa3458fdda0c93fd14cc1ebebe0cbfe703eeL298-R344

FYI Maximilian, please do not feel obligated to backport all these changes in 6.10 branch. I think we can do these once this huge PR is merged together.

Comment (maximiliankolb) [reply guides/common/modules/con_puma-threads.adoc:16]: Thanks; I've added this!
I will backport this for Foreman 3.1, but obviously Satellite doesn't have to use it right away.

**Review by maximiliankolb — COMMENTED** [review]
Rebased to master and applied all feedback.

**Comment by maximiliankolb** [comment]
Still waiting for review.

**Review by jhutar — APPROVED** [review]
Not sure if my review helps to something, but from mine POV this looks great!

**Comment by maximiliankolb** [comment]
Rebased to master once more. I saw no new comments, so I would like to move this ahead and merge this week. Note that we can always fix issues and make things pretty later on. Any objections? cc @Lennonka

**Comment by Lennonka** [comment]
Cherry-picked:

- c9943080..fefe65ce  3.3 -> 3.3
- 9f76ea42..b1c9be52  3.2 -> 3.2
- 301c2463..4f3361da  3.1 -> 3.1

**Pull request merged**
